### PR TITLE
feat: Android multi module kapt

### LINKED ISSUE #1436: android: improve multi modules annotation
Create some strategy to support register widget, actions, and other things in project multi-module.

Examples:
https://developer.android.com/training/dependency-injection/dagger-multi-module

### PR DESCRIPTION
**Pull request by paulomeurerzup**
### Related Issues

Closes #1436 

### Description and Example

This PR aims to change Beagle's Android annotation processor. Now, it will be possible to register Beagle components in a multi-module project.

#### How it gonna work?

From now on, every module or dependency that uses Beagle's annotation processor will generate a group of classes called **_Registrar_**. This classes will keep track of all Beagle components registered inside a specific module.

When generating the `BeagleSetup` class for a module, Beagle's annotation processor will look for **_Registrar_** classes inside a previously known package named **_br.com.zup.beagle.android.registrar_** on the processing environment.

Beagle's annotation processor will then join the currently processing module registered components with the components found on _**Registrar**_ classes of dependencies.

#### Meet Registrar classes

**_Registrar_** classes will be always generated inside a known package named: `br.com.zup.beagle.android.registrar`. This is necessary, because kapt will look for **_Registrar_** classes inside this package on dependency tree every time it process module annotations.

Due the fact that every module will generate **_Registrar_** classes in a package with the same name across the project, **_Registrar_** classes names must have a dynamic suffix. This suffix can be automatically generated, or can be configured using kapt argument in build.gradle file:

```
kapt {
    arguments {
        arg("beagle.moduleName", "sample")
    }
}
```

This configuration will generate **_Registrar_** classes names like: _RegisteredActionsRegistrarSample_.

For sake of readability, it's always recommended to use `beagle.moduleName` argument. If this argument isn’t present, the annotation processor will generate a random UUID suffix for **_Registrar_** classes.

Dynamic generated classes names will look like: _RegisteredActionsRegistrar59c20e58331c4f11b84d377475593739_

Generated classes:

_PropertiesRegistrar_: Keeps track of all BeagleSetup properties registered inside the module.

_RegisteredActionsRegistrar_: Keeps track of all actions registered inside the module.

_RegisteredControllerRegistrar_: Keeps track of all controllers registered inside the module.

_RegisteredCustomTypeAdaptersRegistrar_: Keeps track of all type adapters registered inside the module.

_RegisteredCustomValidatorRegistrar_: Keeps track of all validators registered inside the module.

_RegisteredOperationsRegistrar_: Keeps track of all operations registered inside the module.

_RegisteredWidgetsRegistrar_: Keeps track of all widgets registered inside the module.

#### BeagleSetup and BeagleConfig issue

Before this change, every module that used Beagle's annotation processor generated the `BeagleSetup` class. This is a problem when you want to create a module that, for instance, will have only widgets.

The problem is that `BeagleConfig` is a non null property inside `BeagleSetup` class so, we’ve ended up having to register a dummy `BeagleConfig` class inside that module.

To solve this situation, now it’s possible to disable `BeagleSetup` class generation in modules where it’s not needed. This is done by adding the following kapt argument to the beagle.gradle:

```
kapt {
    arguments {
        arg("beagle.generateSetupClasses", "false")
    }
}
```

#### Properties that are unique

Some properties are unique inside `BeagleSetup` class. This means that only one component of that type can be registered on your entire project structure. Properties that are unique includes:

- DesignSystem;
- HttpClient;
- BeagleConfig;
- …

If you register a component that should be unique twice on your project structure, Beagle’s annotation processor will complain and doesn’t compile your project.

Beagle’s annotation processor was changed to always show the complete path to components that were defined multiple times in the error message. It will look like:

> error: BeagleConfig defined multiple times: 
> public final class AppBeagleConfig implements br.com.zup.beagle.android.setup.BeagleConfig {
>              ^
>   br.com.zup.beagle.sample.AppBeagleConfig
>   br.com.zup.beagle.android.modulea.ModuleABeagleConfig
>   
>   You must remove one implementation from the application.

#### Testing

For helping tests, I've created a repository with the sample project and some modules dependencies:

https://github.com/paulomeurerzup/annotation-test-modules

> To use this sample project, it's necessary to deploy the Beagle annotation processor **from this branch** to mavenLocal.

# TL;DR

Now it’s possible to register Beagle components in a multi-module project. You can have a module where you gonna register your Widgets, other module where you’ll do network configuration and so on.

Every module that will register some Beagle component must define Beagle annotation processor kapt dependency in the build.gradle file e.g.

`kapt "br.com.zup.beagle:android-processor:1.0.0"`

You can specify if a specific module needs to generate `BeagleSetup` class using _**beagle.generateSetupClasses**_ kapt argument and you can specify Registrar classes name suffix (for sake of readability) using _**beagle.moduleName**_ kapt argument e.g.

```
kapt {
    arguments {
        arg("beagle.moduleName", "sample")
        arg("beagle.generateSetupClasses", "false")
    }
}
```

### Checklist

Please, check if these important points are met using `[x]`:

- [X] I read the [PR Guide] and followed the process outlined there for submitting this PR.
- [X] I avoided _breaking changes_ by not changing public APIs that people rely on. 
- [X] I am willing to follow-up on review comments in a timely manner.
- [X] I have made the documentation changes or I created an issue explaining how to document this change on [Docs issues](https://github.com/ZupIT/beagle-docs/issues). Please link the issue here:

[PR Guide]: https://github.com/ZupIT/beagle/blob/master/doc/contributing/pull_requests.md

### CODE REVIEW

**Comment by botbeaglezup** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/BeagleProcessorUtils.kt [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+package br.com.zup.beagle.compiler.shared
+
+import javax.annotation.processing.ProcessingEnvironment
+
+fun beagleClassesGenerationDisabled(processingEnv: ProcessingEnvironment): Boolean {
+    return processingEnv.options.getOrDefault(KAPT_BEAGLE_GENERATE_SETUP_OPTION_NAME, "") == "false"
+}
+
+@Suppress("UNCHECKED_CAST")
+fun forEachRegisteredDependency(
```
Comment (botbeaglezup) [inline android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/BeagleProcessorUtils.kt:26]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/BeagleProcessorUtils.kt#L26"></span>Function forEachRegisteredDependency is nested too deeply.</td>
    </tr>
  </tbody>
</table>

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupProcessor.kt [hunk 2]
```diff
@@ -132,7 +249,26 @@ internal data class BeagleSetupProcessor(
         }
     }
 
+    private fun registrarModuleDefinedProperties(roundEnvironment: RoundEnvironment) {
+        val properties = beagleSetupPropertyGenerator.generate(roundEnvironment, true)
+        PropertiesRegistrarProcessor().process(processingEnv, properties, "$PROPERTIES_REGISTRAR_CLASS_NAME${getModuleName()}")
```
Comment (botbeaglezup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupProcessor.kt:254]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupProcessor.kt#L254"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupPropertyGenerator.kt [hunk 3]
```diff
@@ -34,9 +37,27 @@ internal class BeagleSetupPropertyGenerator(private val processingEnv: Processin
...
+        val propertySpecifications = PropertySpecifications()
 
+        // Get properties possibly registered in dependency modules
+        if (!onlyPropertiesRegisteredInsideModule) {
+            forEachRegisteredDependency(
+                processingEnv,
+                PROPERTIES_REGISTRAR_CLASS_NAME,
+                PROPERTIES_REGISTRAR_METHOD_NAME
+            ) { registeredDependency ->
+                var typeElement = processingEnv.elementUtils.getTypeElement(registeredDependency.second.removeSuffix("()"))
```
Comment (botbeaglezup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupPropertyGenerator.kt:51]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupPropertyGenerator.kt#L51"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupPropertyGenerator.kt [hunk 4]
```diff
@@ -34,9 +37,27 @@ internal class BeagleSetupPropertyGenerator(private val processingEnv: Processin
...
+                if (typeElement == null) {
+                    typeElement = processingEnv.elementUtils.getTypeElement(registeredDependency.second.substringBefore("::"))
```
Comment (botbeaglezup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupPropertyGenerator.kt:53]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupPropertyGenerator.kt#L53"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/PropertiesRegistrarProcessor.kt [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+            |val $PROPERTIES_REGISTRAR_PROPERTY_NAME = listOf<Pair<String, String>>(
+            |   $propertiesRegistered
+            |)
+        |""".trimMargin()
+
+    fun registeredProperties(properties: List<PropertySpec>): String {
+        val registeredProperties = StringBuilder()
+        properties.forEach { propertySpec ->
+            if (!propertySpec.initializer.toString().startsWith("null")) {
+                registeredProperties.append("\tPair(\"${propertySpec.name}\", \"\"\"${propertySpec.initializer.toString()}\"\"\"),\n")
```
Comment (botbeaglezup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/PropertiesRegistrarProcessor.kt:91]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor/src/main/java/br/com/zup/beagle/android/compiler/PropertiesRegistrarProcessor.kt#L91"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/GenerateFunctionComponentRegistrar.kt [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+                ClassName("kotlin", "String"),
+            )
+        )
+
+        return FunSpec.builder(name)
+            .returns(listReturnType)
+    }
+
+    override fun buildCodeByElement(element: Element, annotation: Annotation): String {
+        return "\n\tPair(\"\"\"${buildElementKey(element, annotation)}\"\"\",\"${getPackageName(element)}.${element.simpleName}\"),"
```
Comment (botbeaglezup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/GenerateFunctionComponentRegistrar.kt:56]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/GenerateFunctionComponentRegistrar.kt#L56"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/GenerateFunctionCustomAdapter.kt [hunk 7]
```diff
@@ -91,78 +84,94 @@ class GenerateFunctionCustomAdapter(private val processingEnv: ProcessingEnviron
...
 
-    companion object {
-        const val REGISTERED_CUSTOM_ADAPTER = "getAdapter"
-        const val JAVA_CLASS = "::class.java"
-        const val BEAGLE_TYPE_ADAPTER_INTERFACE = "BeagleTypeAdapter<T>"
-        const val TYPES_INSTANCE = "ParameterizedTypeFactory.new(\n"
-        const val T_GENERIC = "T"
-        const val BREAK_LINE = "\n"
+    override fun buildCodeByDependency(registeredDependency: Pair<String, String>): String {
+        return "${registeredDependency.first} -> ${registeredDependency.second}() as $BEAGLE_TYPE_ADAPTER_INTERFACE$BREAK_LINE"
```
Comment (botbeaglezup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/GenerateFunctionCustomAdapter.kt:174]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/GenerateFunctionCustomAdapter.kt#L174"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/RegisterControllerProcessor.kt [hunk 8]
```diff
@@ -129,16 +135,41 @@ internal class RegisterControllerProcessor(private val processingEnv: Processing
...
+                processingEnv.messager?.error("Default controller defined multiple times: " +
+                    "\n${defaultActivityRegistered.substringBefore(CONTROLLER_DEFINITION_SUFFIX)}" +
+                    "\n${registeredDependency.second}" +
+                    "\n\nYou must remove one implementation from the application.")
+            }
+
+            if (registeredDependency.first.isEmpty()) {
+                defaultActivityRegistered = "${registeredDependency.second}$CONTROLLER_DEFINITION_SUFFIX"
+            } else {
+                registeredWidgets.append("\n    \"${registeredDependency.first}\" -> ${registeredDependency.second}$CONTROLLER_DEFINITION_SUFFIX")
```
Comment (botbeaglezup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/RegisterControllerProcessor.kt:170]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/android/processor/src/main/java/br/com/zup/beagle/android/compiler/generatefunction/RegisterControllerProcessor.kt#L170"></span>Line detected that is longer than the defined maximum line length in the code style.</td>
    </tr>
  </tbody>
</table>

File: backend/sample/core/src/main/kotlin/br/com/zup/beagle/sample/builder/AnnotationTestBuilder.kt [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package br.com.zup.beagle.sample.builder
+
+import br.com.zup.beagle.analytics.ScreenEvent
```
Comment (botbeaglezup) [inline backend/sample/core/src/main/kotlin/br/com/zup/beagle/sample/builder/AnnotationTestBuilder.kt:19]: <table data-meta="generated_by_backend">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/515094e1cfd1c111ab635b3927e2deecba4af364/backend/sample/core/src/main/kotlin/br/com/zup/beagle/sample/builder/AnnotationTestBuilder.kt#L19"></span>The import 'br.com.zup.beagle.analytics.ScreenEvent' is unused.</td>
    </tr>
  </tbody>
</table>

File: android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/RegistrarComponentsProvider.kt [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package br.com.zup.beagle.compiler.shared
+
+import javax.annotation.processing.ProcessingEnvironment
+
+typealias RegisteredComponentId = String
+typealias RegisteredComponentFullName = String
+
+abstract class RegistrarComponentsProvider {
```
Comment (botbeaglezup) [inline android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/RegistrarComponentsProvider.kt:24]: <table data-meta="generated_by_android">
  <tbody>
    <tr>
      <td>:warning:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/ZupIT/beagle/blob/f599efc6e42891267ef2ce0495e97d4082f58082/android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/RegistrarComponentsProvider.kt#L24"></span>An abstract class without a concrete member can be refactored to an interface.</td>
    </tr>
  </tbody>
</table>

File: android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/RegistrarComponentsProvider.kt [hunk 11]
```diff
@@ -0,0 +1,30 @@
...
+ */
+
+package br.com.zup.beagle.compiler.shared
+
+import javax.annotation.processing.ProcessingEnvironment
+
+typealias RegisteredComponentId = String
+typealias RegisteredComponentFullName = String
+
+interface RegistrarComponentsProvider {
```
Comment (igorgiumellizup) [inline android/processor-shared-code/src/main/kotlin/br/com/zup/beagle/compiler/shared/RegistrarComponentsProvider.kt:24]: RegisterComponentsProvider*

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleProcessorUtils.kt [hunk 12]
```diff
@@ -0,0 +1,23 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package br.com.zup.beagle.android.compiler
+
+import javax.annotation.processing.ProcessingEnvironment
+
+fun beagleClassesGenerationDisabled(processingEnv: ProcessingEnvironment): Boolean {
```
Comment (igorgiumellizup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleProcessorUtils.kt:21]: It's not clear what this method does. Try adding a verb to its name, following method naming conventions

File: android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupProcessor.kt [hunk 13]
```diff
@@ -44,43 +56,151 @@ internal data class BeagleSetupProcessor(
...
         processingEnv,
         REGISTERED_OPERATIONS_GENERATED,
-        GenerateFunctionOperation(processingEnv)
+        GenerateFunctionOperation(
+            processingEnv,
+            DependenciesRegistrarComponentsProvider
+        )
+    )
+
+    private val actionRegistrarFactoryProcessor = GenericFactoryProcessor(
```
Comment (igorgiumellizup) [inline android/processor/src/main/java/br/com/zup/beagle/android/compiler/BeagleSetupProcessor.kt:99]: register*

**Review by igorgiumellizup — CHANGES_REQUESTED** [review]
ok

**Comment by igorgiumellizup** [comment]
agora q vi q ta em draft, sorry heheeh

**Review by uziasferreirazup — APPROVED** [review]

**Review by matheusribeirozup — APPROVED** [review]
